org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenImpl.java

```diff
@@ -853,16 +853,19 @@ public String getArtifactPath(ArtifactRepository repository, String groupId, Str
    */
   public boolean isUnavailable(String groupId, String artifactId, String version, String type, String classifier,
       List<ArtifactRepository> remoteRepositories) throws CoreException {
+
     Artifact artifact = lookup(RepositorySystem.class).createArtifactWithClassifier(groupId, artifactId, version, type,
         classifier);
-
     ArtifactRepository localRepository = getLocalRepository();
 
     File artifactFile = new File(localRepository.getBasedir(), localRepository.pathOf(artifact));
 
     if(artifactFile.canRead()) {
       // artifact is available locally
-      return false;
+      Artifact artifactBin = lookup(RepositorySystem.class).createArtifactWithClassifier(groupId, artifactId, version,
```

Comment:
As the PR is updated the discussion is lost: https://github.com/eclipse-m2e/m2e-core/pull/238#discussion_r652737970

> I mean that this line is in general about skipping update if artifact is already available, and your code is about doing the update whenever there is a newer version available remotely.
> This is a good improvement in general, but I don't think it's a requirement for the story of "downloading sources/javadoc with force update".

This change is required. Just the fact of existence of the sources jar file before the change made this code to return false (=="artifact is available locally" == "no need to re-download"), thus skipping the sources jar from being updated.

We must return true here in case we really need to re-download the sources jar (true=="no artifact is available"=="need to re-download"), otherwise it will never be updated (regardless of if it "forced" or not) when its plain jar is updated by maven.

Reply:
OK, I see, but this logic seems misplaced. MavenImpl shouldn't be touched for this work.
It should be in `BuildPathManager.isUnavailable()` or even better a new `BuildPathManager.isOutdatedSnapshot()` method, and then consumers in BuildPathManager would call `if (isUnavailable() || isOutdatedSnapshot())`